File: dynamic_programming/1d/parantheses_factory/momo/parantheses.py
class Stack:
     def __init__(self):
         self.items = []

     def isEmpty(self):
         return self.items == []

     def push(self, item):
         self.items.append(item)

     def pop(self):
         return self.items.pop()

     def peek(self):
         return self.items[len(self.items)-1]

     def size(self):
         return len(self.items)


n = int(input())
input = input()
stack = Stack()
result = [0]*(n+1)
i = n-1

while i != -1:
	if input[i]==')' or input[i]== ']' or input[i]=='>' or input[i]== '}':
		stack.push(input[i])
	elif not (stack.isEmpty()):
		if (input[i] == '(' and stack.peek() == ')' ) \
				or (input[i] == '[' and stack.peek() == ']')\
				or (input[i] == '<' and stack.peek() == '>')\
				or (input[i] == '{' and stack.peek() == '}'):
			stack.pop()
			result[i] = 2
			if i < n-2:
				result[i] += result[i + 1]
				result[i] += result[i + result[i]]
		else:
			while not stack.isEmpty():
				stack.pop()

	i -= 1
for i in range(n):
	print(result[i], end=" ")

File: dynamic_programming/1d/parantheses_factory/momo/parantheses.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class parantheses {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String line1 = br.readLine();
        int n= Integer.parseInt(line1);
        String in = br.readLine();
        int[] result = new int[n+1];
        Stack stack = new Stack();
        for (int i=n-1; i>=0; i--){
            if(in.charAt(i)==')' || in.charAt(i)== ']' || in.charAt(i)=='>' || in.charAt(i)== '}') {
                stack.push(in.charAt(i));
            }
                else if(!(stack.isEmpty())){
                char top = (char) stack.peek();
                        if((in.charAt(i) == '(' && top==')') ||
                                (in.charAt(i) == '[' && top==']') ||
                                (in.charAt(i) == '<' && top=='>') ||
                                (in.charAt(i) == '{' && top=='}')){
                            stack.pop();
                            result[i]=2;
                            if(i<n-2){
                                result[i] += result[i+1];
                                result[i] += result[i+result[i]];
                            }
                }else {
                            stack.clear();
                        }
            }
        }
        for (int i=0;i<n;i++)
            System.out.print(result[i]+ " ");
    }
}
